File: backend/api/geo.py
import math


def haversine_distance_km(point_a: dict, point_b: dict) -> float | None:
    try:
        lat1 = float(point_a["lat"])
        lon1 = float(point_a["lng"])
        lat2 = float(point_b["lat"])
        lon2 = float(point_b["lng"])
    except (KeyError, TypeError, ValueError):
        return None

    earth_radius_km = 6371
    d_lat = math.radians(lat2 - lat1)
    d_lon = math.radians(lon2 - lon1)

    a = (
        math.sin(d_lat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(d_lon / 2) ** 2
    )
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    return earth_radius_km * c

File: backend/server/utils/geo.js
export function toRadians(value) {
  return (Number(value) * Math.PI) / 180
}

export function haversineDistanceKm(pointA, pointB) {
  const lat1 = Number(pointA?.lat)
  const lon1 = Number(pointA?.lng)
  const lat2 = Number(pointB?.lat)
  const lon2 = Number(pointB?.lng)

  if (![lat1, lon1, lat2, lon2].every(Number.isFinite)) {
    return null
  }

  const earthRadiusKm = 6371
  const dLat = toRadians(lat2 - lat1)
  const dLon = toRadians(lon2 - lon1)

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRadians(lat1)) * Math.cos(toRadians(lat2)) * Math.sin(dLon / 2) * Math.sin(dLon / 2)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))

  return earthRadiusKm * c
}

